pkg/filemanager/filemanager.go

```diff
@@ -44,13 +46,91 @@ func NewFileManager(path string) (FileManager, error) {
 		return NewGCSStorageFile(path)
 	case strings.Contains(path, "blob.core.windows.net"):
 		return NewAzureBlobFile(path)
+	case strings.HasPrefix(path, "http://"), strings.HasPrefix(path, "https://"):
+		return NewCustomS3File(path)
 	case path == "":
 		return nil, errors.New("empty path")
 	default:
 		return NewSystemFile(path), nil
 	}
 }
 
+type CustomS3File struct {
+	s3Client *s3.Client
+	bucket   string
+	key      string
+}
+
+func NewCustomS3File(path string) (*CustomS3File, error) {
+	u, err := url.Parse(path)
+	if err != nil {
+		return nil, err
+	}
+
+	clPath := gp.Clean(u.Path)
+
+	if len(strings.Split(clPath, "/")) < 3 {
+		return nil, fmt.Errorf("invalid s3 path: %s", path)
+	}
+
+	// Extract bucket and path from url
+	bucket, key, _ := strings.Cut(strings.TrimLeft(clPath, "/"), "/")
+
+	if bucket == "" || key == "" {
+		return nil, fmt.Errorf("invalid s3 path: %s", path)
+	}
+
+	const defaultRegion = "us-east-1"
+
+	resolver := aws.EndpointResolverWithOptionsFunc(func(service, region string, options ...any) (aws.Endpoint, error) {
+		return aws.Endpoint{
+			PartitionID:       "aws",
+			SigningRegion:     defaultRegion,
+			URL:               fmt.Sprintf("%v://%v", u.Scheme, u.Host),
+			HostnameImmutable: true,
+		}, nil
+	})
+
+	cfg, err := config.LoadDefaultConfig(
+		context.Background(),
+		config.WithDefaultRegion(defaultRegion),
+		config.WithEndpointResolverWithOptions(resolver))
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &CustomS3File{
+		s3Client: s3.NewFromConfig(cfg),
+		bucket:   bucket, // bucket
+		key:      key,    // path/to/file.csv
+	}, nil
```

Comment:
Can you just use `S3File{}` here?

Reply:
> Can you just return `S3File{}` here? I don't see the difference.

You're right, my initial thought was to keep the original AWS S3 and the custom S3 code completely separated and have each their own Upload and Download implementation. But in retrospect it's probably better to reuse the code.

Reply:
I can see arguments for having it separate as well. 
I don't have a strong opinion on this. I think either way is fine.